crepecake/src/main/java/com/google/cloud/tools/crepecake/cache/CacheWriter.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2018 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.crepecake.cache;
+
+import com.google.cloud.tools.crepecake.blob.BlobDescriptor;
+import com.google.cloud.tools.crepecake.hash.CountingDigestOutputStream;
+import com.google.cloud.tools.crepecake.image.DescriptorDigest;
+import com.google.cloud.tools.crepecake.image.UnwrittenLayer;
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.zip.GZIPOutputStream;
+
+/** Writes {@link UnwrittenLayer}s to the cache. */
+public class CacheWriter {
+
+  private final Cache cache;
+
+  public CacheWriter(Cache cache) {
+    this.cache = cache;
+  }
+
+  public CachedLayer writeLayer(UnwrittenLayer layer) throws IOException {
+    // Writes to a temporary file first because the UnwrittenLayer needs to be written first to
+    // obtain its digest.
+    File tempLayerFile = Files.createTempFile(cache.getCacheDirectory(), null, null).toFile();
```

Comment:
If we delete on exit, does that not mess with what happens after renaming?

Reply:
`deleteOnExit` called `DeleteOnExitHook.add(path)`, which adds the file path to a shutdown hook. When the hook is activated, the file is deleted with `(new File(filename)).delete();`, which doesn't do anything if the `filename` doesn't exist anymore, which is the case here after renaming.